deltachat-ios/Chat/AppPickerViewController.swift

```diff
@@ -0,0 +1,153 @@
+        webView.translatesAutoresizingMaskIntoConstraints = false
+        webView.load(URLRequest(url: url))
+
+        downloadingView = DownloadingView()
+        downloadingView.translatesAutoresizingMaskIntoConstraints = false
+        downloadingView.isHidden = true
+
+        super.init(nibName: nil, bundle: nil)
+
+        webView.navigationDelegate = self
+        view.addSubview(webView)
+        view.addSubview(downloadingView)
+        view.backgroundColor = .systemBackground
+        setupConstraints()
+        let closeButton = UIBarButtonItem(image: UIImage(systemName: "xmark"), style: .plain, target: self, action: #selector(AppPickerViewController.close(_:)))
+
+        title = String.localized("webxdc_apps")
+        navigationItem.leftBarButtonItem = closeButton
+        self.defaultCloseButton = closeButton
+    }
+
+    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
+
+    private func setupConstraints() {
+        let constraints = [
+            webView.topAnchor.constraint(equalTo: view.topAnchor),
+            webView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            view.trailingAnchor.constraint(equalTo: webView.trailingAnchor),
+            view.bottomAnchor.constraint(equalTo: webView.bottomAnchor),
+
+            downloadingView.topAnchor.constraint(equalTo: view.topAnchor),
+            downloadingView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            view.trailingAnchor.constraint(equalTo: downloadingView.trailingAnchor),
+            view.bottomAnchor.constraint(equalTo: downloadingView.bottomAnchor),
+        ]
+
+        NSLayoutConstraint.activate(constraints)
+    }
+
+    // MARK: - Actions
+
+    @objc func close(_ sender: Any) {
+        dismiss(animated: true)
+    }
+
+    @objc func showLoading() {
+        title = String.localized("Downloading...")
+        downloadingView.isHidden = false
+        downloadingView.activityIndicator.startAnimating()
+        downloadingView.activityIndicator.hidesWhenStopped = true
+        navigationItem.leftBarButtonItem = nil
+    }
+
+    @objc func hideLoading() {
+        title = String.localized("webxdc_apps")
+        downloadingView.isHidden = true
+        downloadingView.activityIndicator.stopAnimating()
+        navigationItem.leftBarButtonItem = defaultCloseButton
+    }
+}
+
+extension AppPickerViewController: WKNavigationDelegate {
+    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping @MainActor (WKNavigationActionPolicy) -> Void) {
+        guard let url = navigationAction.request.url else {
+            return decisionHandler(.cancel)
+        }
+
+        // if url ends with .xdc -> download and store in core and call delegate
+        if url.pathExtension == "xdc" {
+            Task { [weak self] in
+                guard let self else { return }
+                // show spinner instead of close-button
```

Comment:
not blocking, but in general, it'd be better to leave close-button as is UX-wise. download may be slow or whatnot, and the user may want to close the view